src/com/sun/jna/CallbackReference.java

```diff
@@ -236,12 +239,13 @@ private CallbackReference(Callback callback, int callingConvention, boolean dire
             }
             int flags = callback instanceof DLLCallback
                 ? Native.CB_OPTION_IN_DLL : 0;
-            long peer = Native.createNativeCallback(proxy, PROXY_CALLBACK_METHOD,  
-                                                    nativeParamTypes, returnType,
-                                                    callingConvention, flags,
-                                                    encoding);
-            cbstruct = peer != 0 ? new Pointer(peer) : null;
+            peer = Native.createNativeCallback(proxy, PROXY_CALLBACK_METHOD,  
+                                               nativeParamTypes, returnType,
+                                               callingConvention, flags,
+                                               encoding);
         }
+        cbstruct = peer != 0 ? new Pointer(peer) : null;
+        allocatedMemory.put(this, new WeakReference(this));
```

Comment:
For some reason this code was previously only tracking direct-mapped callbacks.  I don't think that was intentional, so moving it here to consistently track _all_ allocated callback structures.